repo2docker/contentproviders/base.py

```diff
@@ -0,0 +1,58 @@
+"""
+Base classes for repo2docker ContentProviders
+
+ContentProviders accept a `spec` of various kinds, and
+provide the contents from the spec to a given output directory.
+"""
+import os
+
+
+class ContentProviderException(Exception):
+    """Exception raised when a ContentProvider can not provide content."""
+    pass
+
+
+class ContentProvider:
+    def __init__(self, logger):
+        self.log = logger
+
+    def detect(self, repo, ref=None, extra_args=None):
+        """Determine compatibility between source and this provider.
+
+        If the provider knows how to fetch this source it will return a
+        `spec` that can be passed to `fetch`. The arguments are the `repo`
+        string passed on the command-line, the value of the --ref parameter,
+        if provided and any provider specific arguments provided on the
+        command-line.
```

Comment:
This `--provider-extra-args` flag doesn't exist yet. My thinking is that future providers might require wildly different kinds of argument which we can't all explicitly support so instead we provide this as a provider specific/internal escape hatch.